File: rs/day5/src/main.rs
use std::env;
use std::error;
use std::fs;

fn run(prog: &intcode::Program, input: i64) -> Result<i64, String> {
    let mut machine = intcode::Machine::new(&prog);
    let mut state = machine.run()?;
    let mut output = 0;
    loop {
        if state == intcode::State::Reading {
            machine.write(input);
        } else if state == intcode::State::Writing {
            output = machine.read();
        }
        state = machine.run()?;
        if state == intcode::State::Halted {
            break;
        }
        if output != 0 {
            return Err(format!("diagnostic failed: {}", output));
        }
    }
    Ok(output)
}

fn main() -> Result<(), Box<dyn error::Error>> {
    let path = env::args().nth(1).ok_or("missing input path")?;
    let text = fs::read_to_string(&path)?;
    let prog = intcode::Program::new(&text)?;
    println!("{}", run(&prog, 1)?);
    Ok(())
}

File: js/day5/main.js
"use strict";

const fs = require("fs");
const intcode = require("../intcode");

function run(prog, testId) {
    let diagCode;
    const vm = new intcode.VM(prog);
    while (vm.state !== intcode.State.HALT) {
        switch (vm.state) {
            case intcode.State.RUN:
                vm.run();
                break;
            case intcode.State.WRITE:
                diagCode = vm.read();
                break;
            case intcode.State.READ:
                vm.write(testId);
                break;
        }
    }
    return diagCode;
}

function main(args) {
    const path = args[0];
    const file = fs.readFileSync(path, "ascii");
    const toks = file.split(",");
    const prog = toks.map((s) => parseInt(s, 10));
    console.log(run(prog, 1));
    console.log(run(prog, 5));
}

main(process.argv.slice(2));
